// lodash/pad.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
export interface CustomDimensionsAndMetricsOptions {
    dimensions?: (boolean | number | string)[];
    metrics?: (boolean | number | string)[];
}
export interface EventOptions extends CustomDimensionsAndMetricsOptions {
    label?: string;
    value?: string;
}
export interface ScreenviewOptions extends CustomDimensionsAndMetricsOptions {
    appVersion?: string;
    appId?: string;
    appInstallerId?: string;
}
export interface PageviewOptions extends CustomDimensionsAndMetricsOptions {
    hostname?: string;
    title?: string;
}
export interface TimingOptions extends CustomDimensionsAndMetricsOptions {
    label?: string;
}
/**
 * Interface for managing analytics. This is highly platform dependent, and mostly matches
 * Google Analytics. The reason the interface is here is to remove the dependency to an
 * implementation from most other places.
 *
 * The methods exported from this interface more or less match those needed by us in the
 * universal analytics package, see https://unpkg.com/@types/universal-analytics@0.4.2/index.d.ts
 * for typings. We mostly named ar
