// rxjs/_esm2015/internal/AsyncSubject.js
import { AjaxRequest } from './types';
/**
 * A normalized AJAX error.
 *
 * @see {@link ajax}
 *
 * @class AjaxError
 */
export interface AjaxError extends Error {
    /**
     * The XHR instance associated with the error.
     */
    xhr: XMLHttpRequest;
    /**
     * The AjaxRequest associated with the error.
     */
    request: AjaxRequest;
    /**
     * The HTTP status code, if the request has completed. If not,
     * it is set to `0`.
     */
    status: number;
    /**
     * The responseType (e.g. 'json', 'arraybuffer', or 'xml').
     */
    responseType: XMLHttpRequestResponseType;
    /**
     * The response data.
     */
    response: any;
}
export interface AjaxErrorCtor {
    /**
     * @deprecated Internal implementation detail. Do not construct error instances.
     * Cannot be tagged as internal: https://github.com/ReactiveX/rxjs/issues/6269
     */
    new (message: string, xhr: XMLHttpRequest, request: AjaxRequest): AjaxError;
}
/**
 * Thrown when an error occurs during an AJAX request.
 * This is only exported because it is useful for checking to see if
