// rxjs/internal/util/subscribeTo.js
export interface Schema {
    $schema?: string;
    cli?: CliOptions;
    /**
     * Default project name used in commands.
     * @deprecated The project to use will be determined from the current working directory.
     */
    defaultProject?: string;
    /**
     * Path where new projects will be created.
     */
    newProjectRoot?: string;
    projects?: Projects;
    schematics?: SchematicOptions;
    version: number;
}
export interface CliOptions {
    /**
     * Share anonymous usage data with the Angular Team at Google.
     */
    analytics?: Analytics;
    analyticsSharing?: AnalyticsSharing;
    /**
     * Control disk cache.
     */
    cache?: Cache;
    /**
     * The default schematics collection to use.
     */
    defaultCollection?: string;
    /**
     * Specify which package manager tool to use.
     */
    packageManager?: PackageManager;
    /**
     * The list of schematic collections to use.
     */
    schematicCollections?: string[];
    /**
     * Control CLI specific console warnings
     */
    warnings?: Warnings;
}
/**
 * Share anonymous usage data with the Angular Team at Google.
 */
export declare type Analytics = boolean | string;
export interface AnalyticsSharing {
    /**
     * Analytics sharing info tracking ID.
     */
    tracking?: string;
    /**
     * Analytics sharing info universally unique identifier.
     */
    uuid?: string;
}
/**
 * Control disk cache.
 */
export interface Cache {
    /**
